### ISSUE TITLE
UVLoop closes the connection but twisted does not call connectionLost

### ISSUE THREAD

**Issue by r00ta** [twisted/twisted#11904]
Hi there!

I'm using twisted with uvloop as event loop. In particular, when my application is under heavy load I've faced the following exception

```
File "/usr/lib/python3/dist-packages/twisted/internet/asyncioreactor.py", line 173, in addWriter
	self._asyncioEventloop.add_writer(fd, callWithLogger, writer,
File "uvloop/loop.pyx", line 2399, in uvloop.loop.Loop.add_writer
File "uvloop/loop.pyx", line 808, in uvloop.loop.Loop._add_writer
File "uvloop/handles/poll.pyx", line 122, in uvloop.loop.UVPoll.start_writing
File "uvloop/handles/poll.pyx", line 39, in uvloop.loop.UVPoll._poll_start
File "uvloop/handles/handle.pyx", line 159, in uvloop.loop.UVHandle._ensure_alive
builtins.RuntimeError: unable to perform operation on <UVPoll closed=True 0x7fe32f0f9cf0>; the handler is closed
```

which is fair so far. The problem is that the `connectionLost` callback is not being called, preventing my application to actually detect that the connection was closed and that it should not be used anymore - as you can imagine, the application ended up in an infinite loop trying to use such closed connection. 

The application looks like 

```
import asyncio
import uvloop

asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
print(f"Event loop implementation: {str(type(asyncio.get_event_loop_policy()))}")

from twisted.internet import asyncioreactor
asyncioreactor.install()

class MyAMPFactory(protocol.Factory):
  ...

reactor = AsyncioSelectorReactor()
tcp = reactor.listenTCP(12345, MyAMPFactory())
...
```

Unfortunately I don't have a plain reproducer as I faced this issue on some environments under heavy load. I was able to patch the uvloop library, close the connection (like an exception from the underlying libuv loop would do) and assess that the `connectionLost` callback was not fired.
You can find some detaiils in the bug I've opened on [uvloop side](https://github.com/MagicStack/uvloop/issues/552) (as I'm new to the twisted/uvloop internals, I don't know if it's a bug on uvloop side or on twisted side). But looking at the exceptions being catched [here](https://github.com/twisted/twisted/blob/a9ee8e59a5cd1950bf1a18c4b6ca813e5f56ad08/src/twisted/internet/asyncioreactor.py#L175), I suspect this case should be handled by twisted accordingly.  
 

**Describe the correct behavior you'd like to see**
If the connection is closed on uvloop side, the `connectionLost` callback should be triggered as well.

**Testing environment**
 - Operating System and Version
 ```
Linux focal 5.17.0-1032-oem #33-Ubuntu SMP PREEMPT Thu May 18 15:34:21 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.6 LTS"
```
 - Twisted version: 18.9.0
 - Reactor: uvloo

**Comment by adiroiban** [comment]
Thanks Jacopo for the report.

Can you try to reproduce this using latest Twisted release?

I guess you are using the Ubuntu default python 3.7 version.

The current Twisted contributors team is small and we don't have the power to support older Twisted releases ... especially one released almost 5 years ago

**Comment by r00ta** [comment]
> Thanks Jacopo for the report.
> 
> Can you try to reproduce this using latest Twisted release?
> 
> I guess you are using the Ubuntu default python 3.7 version.
> 
> The current Twisted contributors team is small and we don't have the power to support older Twisted releases ... especially one released almost 5 years ago

Sure, I understand. I'll try to reproduce this with the latest one and get back to you! Thanks :+1:

**Comment by glyph** [comment]
@r00ta any luck? I really appreciate the detailed report but I want to make sure we don't leave this in an un-actionable state.

**Comment by r00ta** [comment]
Hey @glyph, I was about to report that I've found the issue and it was basically in our codebase. 

In short, I found out that twisted called [unhandledError](https://github.com/twisted/twisted/blob/a9ee8e59a5cd1950bf1a18c4b6ca813e5f56ad08/src/twisted/protocols/amp.py#L2495) originally but we had an overwrite of that method in our codebase - and we were not calling `loseConnection`. This is how we ended up with a zombie connection. 

I'm closing this!
